File: src/main/java/implementslegendkt/mod/vaultjp/JewelPurposerBlockEntity.java
package implementslegendkt.mod.vaultjp;

import implementslegendkt.mod.vaultjp.network.Channel;
import implementslegendkt.mod.vaultjp.network.UpdatePurposesPacket;
import iskallia.vault.block.entity.VaultJewelApplicationStationTileEntity;
import iskallia.vault.block.entity.VaultJewelCuttingStationTileEntity;
import iskallia.vault.config.VaultJewelCuttingConfig;
import iskallia.vault.container.VaultJewelApplicationStationContainer;
import iskallia.vault.container.oversized.OverSizedInventory;
import iskallia.vault.gear.attribute.VaultGearAttributeInstance;
import iskallia.vault.gear.attribute.type.VaultGearAttributeTypeMerger;
import iskallia.vault.gear.data.ToolGearData;
import iskallia.vault.gear.data.VaultGearData;
import iskallia.vault.gear.item.VaultGearItem;
import iskallia.vault.init.ModAttributes;
import iskallia.vault.init.ModConfigs;
import iskallia.vault.init.ModGearAttributes;
import iskallia.vault.item.tool.JewelItem;
import iskallia.vault.item.tool.ToolItem;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.world.Container;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.WorldlyContainer;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.entity.TickingBlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.items.CapabilityItemHandler;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.wrapper.InvWrapper;
import net.minecraftforge.items.wrapper.SidedInvWrapper;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;

public class JewelPurposerBlockEntity extends BlockEntity implements MenuProvider {

    public static final BlockEntityType<JewelPurposerBlockEntity> TYPE = BlockEntityType.Builder.of(JewelPurposerBlockEntity::new, JewelPurposerBlock.INSTANCE).build(null);
    private final JewelPurposerInventory inventory = new JewelPurposerInventory();

    public ArrayList<JewelPurpose> purposes = new ArrayList<>();

    private LazyOptional<? extends IItemHandler> handler = LazyOptional.of(()->new InvWrapper(inventory));
    public JewelPurposerBlockEntity(BlockPos p_155229_, BlockState p_155230_) {

        super(TYPE, p_155229_, p_155230_);
    }

    @Override
    public Component getDisplayName() {
        return new TextComponent("Jewel Purposer");
    }

    @Nullable
    @Override
    public AbstractContainerMenu createMenu(int p_39954_, Inventory p_39955_, Player p_39956_) {
        return this.getLevel() == null ? null : new JewelPurposerContainer(p_39954_, this.getLevel(), this.getBlockPos(), p_39956_.getInventory());
    }

    @NotNull
    @Override
    public <T> LazyOptional<T> getCapability(@NotNull Capability<T> cap) {
        if (!this.remove && cap == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) return this.handler.cast();

        return super.getCapability(cap);
    }

    @Override
    public void invalidateCaps() {
        super.invalidateCaps();
        handler.invalidate();
    }

    @Override
    public void reviveCaps() {
        super.reviveCaps();
        this.handler = LazyOptional.of(()->new InvWrapper(inventory));
    }


    @Override
    public void load(CompoundTag tag) {
        super.load(tag);
        for (var purposeTag:tag.getList("purposes",CompoundTag.TAG_COMPOUND)) {
            purposes.add(JewelPurpose.readNBT((CompoundTag) purposeTag));
        }
        this.inventory.load(tag);
    }

    @Override
    public void saveAdditional(CompoundTag tag) {
        super.saveAdditional(tag);
        var purposesTag = new ListTag();
        for (var purpose :purposes) {
            purposesTag.add(JewelPurpose.writeNBT(purpose));
        }
        tag.put("purposes",purposesTag);
        this.inventory.save(tag);
    }


    public CompoundTag getUpdateTag() {
        return this.saveWithoutMetadata();
    }

    @javax.annotation.Nullable
    public ClientboundBlockEntityDataPacket getUpdatePacket() {
        return ClientboundBlockEntityDataPacket.create(this);
    }

    public JewelPurposerInventory getInventory() {
        return this.inventory;
    }


    public boolean stillValid(Player player) {
        return this.level != null && this.level.getBlockEntity(this.worldPosition) == this && this.inventory.stillValid(player);
    }


    public ItemStack applyJewelsMock(ItemStack tool, int[] jewels) {
        for (var jew : jewels) {
            if(jew<0 || jew>JewelPurposerContainer.JEWEL_COUNT_MAX)break;
            VaultJewelApplicationStationTileEntity.applyJewel(tool,inventory.getItem(jew));
        }
        return tool;
    }
    public void applyJewels(int[] jewels) {
        var toolIndex = inventory.getContainerSize()-1;
        var toolOld = inventory.removeItemNoUpdate(toolIndex);
        ItemStack tool;
        if(!(toolOld.getItem() instanceof ToolItem)){
            inventory.setItem(toolIndex,toolOld);
            return;
        }

        for (var jew : jewels) {
            if(jew>=0 && jew<JewelPurposerContainer.JEWEL_COUNT_MAX) {
                if (ToolGearData.read(toolOld).get(ModGearAttributes.TOOL_CAPACITY, VaultGearAttributeTypeMerger.intSum()) <10)break;
                tool=toolOld.copy();
                var jewel = inventory.removeItemNoUpdate(jew);
                VaultJewelApplicationStationTileEntity.applyJewel(tool,jewel);
                if(ToolGearData.read(tool).get(ModGearAttributes.TOOL_CAPACITY, VaultGearAttributeTypeMerger.intSum()) <0){
                    inventory.setItem(jew,jewel);
                }else {
                    toolOld=tool;
                }
            }
        }

        inventory.setItem(toolIndex,toolOld);

    }

    private int mod = 0;
    public void tick() {
        if(level.isClientSide())return;
        mod= (mod+1)%16;
        var range = ModConfigs.VAULT_JEWEL_CUTTING_CONFIG.getJewelCuttingRange();
        for (int i = mod; i < JewelPurposerContainer.JEWEL_COUNT_MAX; i+=16) {
            if (inventory.getItem(i).isEmpty()) continue;
            var item = inventory.removeItemNoUpdate(i);
            if (!(item.getItem() instanceof JewelItem)) {
                inventory.setItem(i, item);
                continue;
            }
            var tag = item.getOrCreateTag();
            var cuts = tag.getInt("freeCuts");
            if (cuts < 3) {
                tag=tag.copy();
                var data = VaultGearData.read(item);
                var jewelSize = data.getModifiers(ModGearAttributes.JEWEL_SIZE, VaultGearData.Type.ALL);
                for (var cuts_ = cuts; cuts_ < 3 && jewelSize.stream().anyMatch((it) -> it.getValue() > 10); cuts_++) {
                    for (var sizeInstance : jewelSize) {
                        sizeInstance.setValue(Integer.max(10, sizeInstance.getValue() - range.getRandom()));
                    }
                }
                tag.putInt("freeCuts", 3);
                item.setTag(tag);
                data.write(item);
            }
            inventory.setItem(i,item);
            if (cuts < 3)tryRecycleJewel(i);

        }
    }
    private void tryRecycleJewel(int slot){
        var jewel = inventory.getItem(slot);
        if(purposes.isEmpty() || slot<0 || slot>= JewelPurposerContainer.JEWEL_COUNT_MAX){
            return;
        }
        for (var purpose :
                purposes) {
            var usefulness = purpose.getJewelUsefulness(jewel);
            if (usefulness>purpose.disposeThreshold())return;
        }
        inventory.removeItemNoUpdate(slot);
        inventory.setChanged();
    }

    public void disposeBad() {
        if(purposes.isEmpty()){
            return;
        }
        for (int i = 0; i <JewelPurposerContainer.JEWEL_COUNT_MAX; i++) {
            tryRecycleJewel(i);
        }
    }

    public void syncToServer() {
        Channel.CHANNEL.sendToServer(new UpdatePurposesPacket(getBlockPos(),purposes));
    }


    public class JewelPurposerInventory extends OverSizedInventory implements DefaultWorldlyContainer {
        public JewelPurposerInventory() {super(JewelPurposerContainer.JEWEL_COUNT_MAX + 16+1, JewelPurposerBlockEntity.this);}

    }

}

File: src/main/java/implementslegendkt/mod/vaultjp/JewelPurposerBlockEntity.kt
package implementslegendkt.mod.vaultjp

import implementslegendkt.mod.vaultjp.PurposerExternalStorages.getOrCreateExternalStorage
import implementslegendkt.mod.vaultjp.PurposerExternalStorages.saveExternalStorage
import implementslegendkt.mod.vaultjp.mixin.AccessorDataStorage
import implementslegendkt.mod.vaultjp.network.Channel
import implementslegendkt.mod.vaultjp.network.UpdatePurposesPacket
import iskallia.vault.container.oversized.OverSizedInventory
import iskallia.vault.gear.attribute.type.VaultGearAttributeTypeMerger
import iskallia.vault.gear.data.ToolGearData
import iskallia.vault.init.ModGearAttributes
import iskallia.vault.item.JewelPouchItem
import iskallia.vault.item.tool.JewelItem
import iskallia.vault.item.tool.ToolItem
import iskallia.vault.util.LootInitialization
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.TextComponent
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.MenuProvider
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.util.LazyOptional
import net.minecraftforge.items.CapabilityItemHandler
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.items.wrapper.InvWrapper
import java.util.*

class JewelPurposerBlockEntity(p_155229_: BlockPos?, p_155230_: BlockState?) : BlockEntity(TYPE, p_155229_, p_155230_),
    MenuProvider {


    private var data: PurposerExternalStorage? = null

    private var handler: LazyOptional<out IItemHandler> = LazyOptional.of { InvWrapper(getOrCreateData().inventory) }

    val inventory get() = getOrCreateData().inventory
    val purposes get() = getOrCreateData().purposes

    var externalUUID= UUID.randomUUID()
    
    var oldTag:CompoundTag? = null

    private fun getOrCreateData(): PurposerExternalStorage{
        val _oldTag = oldTag
        oldTag=null
        return data?:run {
            val result = ((level as? ServerLevel)?.let {
                getOrCreateExternalStorage(
                    it.server.overworld().getDataStorage(),
                    externalUUID,
                    this
                )
            } ?: PurposerExternalStorage(externalUUID, this)).also { data = it }

            if(_oldTag!==null){
                result.inventory.load(_oldTag)
                result.purposes.addAll(
                    _oldTag.getList("purposes", CompoundTag.TAG_COMPOUND.toInt())
                        .map { JewelPurpose.readNBT(it as CompoundTag) })

            }
            result
        }
    }

    override fun getDisplayName(): Component = TextComponent("Jewel Purposer")

    override fun createMenu(p_39954_: Int, p_39955_: Inventory, p_39956_: Player): AbstractContainerMenu? =
        this.getLevel()?.let {
            JewelPurposerContainer(
                p_39954_,
                it,
                this.blockPos,
                p_39956_.inventory
            )
        }

    override fun <T> getCapability(cap: Capability<T>, side: Direction?): LazyOptional<T> =
        if (!this.remove && cap === CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) handler.cast()
            else super.getCapability(cap, side)

    override fun invalidateCaps() {
        super.invalidateCaps()
        handler.invalidate()
    }

    override fun reviveCaps() {
        super.reviveCaps()
        this.handler = LazyOptional.of { InvWrapper(inventory) }
    }


    override fun load(tag: CompoundTag) {
        super.load(tag)

        if (tag.hasUUID("externalUUID")) externalUUID = tag.getUUID("externalUUID")
        else oldTag=tag
        //inventory.load(tag)
    }

    public override fun saveAdditional(tag: CompoundTag) {
        super.saveAdditional(tag)
        val _data = data
        tag.putUUID("externalUUID", externalUUID)
        if (_data != null) (level as? ServerLevel)?.let {
            saveExternalStorage(it.dataStorage,_data)
        }
        //inventory.save(tag)
    }


    override fun getUpdateTag(): CompoundTag = this.saveWithoutMetadata()

    override fun getUpdatePacket(): ClientboundBlockEntityDataPacket? = ClientboundBlockEntityDataPacket.create(this)


    fun stillValid(player: Player?): Boolean =
        this.level != null &&
                (level!!.getBlockEntity(this.worldPosition) === this) &&
                inventory.stillValid(player)


    fun applyJewelsMock(tool: ItemStack, jewels: IntArray) = jewels.fold(tool){
        toolAcc,jewelIndex->
        if(jewelIndex !in 0 until JewelPurposerContainer.JEWEL_COUNT_MAX) return@applyJewelsMock toolAcc
        ToolItem.applyJewel(toolAcc,inventory.getItem(jewelIndex), false)
        toolAcc
    }

    fun applyJewels(jewels: IntArray) {
        val toolIndex = inventory.containerSize - 1
        var toolOld = inventory.removeItemNoUpdate(toolIndex)
        var tool: ItemStack
        if (toolOld.item !is ToolItem) {
            inventory.setItem(toolIndex, toolOld)
            return
        }

        for (jew in jewels) {
            if (jew >= 0 && jew < JewelPurposerContainer.JEWEL_COUNT_MAX) {
                if (ToolGearData.read(toolOld)
                        .get(ModGearAttributes.TOOL_CAPACITY, VaultGearAttributeTypeMerger.intSum()) < 10
                ) break
                tool = toolOld.copy()
                val jewel = inventory.removeItemNoUpdate(jew)

                if (ToolItem.applyJewel(tool,jewel,true) && ToolGearData.read(tool)
                        .get<Int, Int>(ModGearAttributes.TOOL_CAPACITY, VaultGearAttributeTypeMerger.intSum()) < 0
                ) {
                    inventory.setItem(jew, jewel)
                } else {
                    toolOld = tool
                }
            }
        }

        inventory.setItem(toolIndex, toolOld)
    }

    fun tick() {
        if (level!!.isClientSide()) return
        disposeBad()
    }

    private fun tryRecycleJewel(slot: Int, cap: IItemHandler): Boolean {
        if (purposes.isEmpty() || slot < 0 || slot >= JewelPurposerContainer.JEWEL_COUNT_MAX) return false
        val jewel = inventory.getItem(slot)
        if (purposes.any { !it.isBad(jewel) } || jewel.item !is JewelItem) return true

        var stack = inventory.getItem(slot)
        for (targetSlot in 0 until cap.slots) {
            if (stack.isEmpty) break
            stack = cap.insertItem(targetSlot, stack, false)
        }
        if (stack.isEmpty) { //assuming 1 item per slot
            inventory.removeItemNoUpdate(slot)
            inventory.setChanged()
        } else return false
        return true
    }

    fun disposeBad() {
        if (purposes.isEmpty()) return

        val next = level?.getBlockEntity(worldPosition.below()) ?: return
        val cap = next.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, Direction.UP)
            .orElse(next.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY).orElse(null))
            ?: return

        repeat (JewelPurposerContainer.JEWEL_COUNT_MAX) {
            if (!tryRecycleJewel(it, cap)) return
        }
    }

    fun syncToServer() = Channel.CHANNEL.sendToServer(UpdatePurposesPacket(blockPos, purposes))

    fun deleteExternalData() {
        data?.deleted=true
        ((level as? ServerLevel)?.dataStorage as? AccessorDataStorage)?.callGetDataFile(PurposerExternalStorages.nameFor(externalUUID))?.delete()
    }

    fun acceptBest(options: List<JewelPouchItem.RolledJewel>,level:Int,dbgPlayer:ServerPlayer?=null) {
        (this.level as? ServerLevel)?.let {
            world->
            val identified = options.filter { it.identified }.map { it.stack }
            val unidentified = options.filter { !it.identified }.map { LootInitialization.initializeVaultLoot(it.stack, level)/*todo identify*/ }.toMutableList()

                    val chosen = (identified.mapNotNull {
                    jewel->
                val ratios = purposes.map {
                    it.getJewelUsefulness(jewel)/it.disposeThreshold
                }
                val highestRatio = ratios.maxOrNull()?:run { unidentified+=jewel;return@mapNotNull null }
                (jewel to highestRatio).takeIf { it.second>1.0 }
            }.maxByOrNull {
                it.second
            }?.first?:unidentified.random()).copy()

            val slot = run slot@{
                repeat(JewelPurposerContainer.JEWEL_COUNT_MAX) {
                    if(inventory.getItem(it).isEmpty)return@slot it
                }
                -1
            }
            if(slot!=-1){
                inventory.setItem(slot,chosen)
            }else{
                world.addFreshEntity(ItemEntity(world, worldPosition.x.toDouble(),
                    worldPosition.y.toDouble(), worldPosition.z.toDouble(),chosen))
            }
        }
    }


    companion object {
        @JvmField
        val TYPE: BlockEntityType<JewelPurposerBlockEntity> = BlockEntityType.Builder.of(
            { pos, state -> JewelPurposerBlockEntity(pos, state) },
            JewelPurposerBlock
        ).build(null)
    }
}


class JewelPurposerInventory(val purposer:JewelPurposerBlockEntity) :
    OverSizedInventory(JewelPurposerContainer.JEWEL_COUNT_MAX + 16 + 1, purposer),
    DefaultWorldlyContainer
{
    override fun canTakeItemThroughFace(p_19235_: Int, stack: ItemStack, directio: Direction): Boolean =
        if (stack.item is JewelItem && purposer.purposes.any { purpose -> purpose.getJewelUsefulness(stack) >= purpose.disposeThreshold }) {
            false
        } else super.canPlaceItemThroughFace(p_19235_, stack, directio)
}
